tools/esdlcmd/esdlcmd_diff.cpp

```diff
@@ -0,0 +1,1139 @@
+        }
+
+        Owned<IFile> file = createIFile(path.str());
+        Owned<IFileIO> io;
+        io.setown(file->open(IFOcreaterw));
+
+        DBGLOG("Writing to file %s", file->queryFilename());
+
+        if (io.get())
+            io->write(0, strlen(text), text);
+        else
+            DBGLOG("File %s can't be created", file->queryFilename());
+    }
+
+public:
+    EsdlCmdHelper cmdHelper;
+
+    StringAttr optService;
+    StringAttr optMethod;
+};
+
+class EsdlDiffGenCmd : public EsdlConvertCmd
+{
+public:
+    EsdlDiffGenCmd() : optFlags(DEPFLAG_COLLAPSE|DEPFLAG_ARRAYOF){}
+
+    virtual bool parseCommandLineOptions(ArgvIterator &iter)
+    {
+        if (iter.done())
+        {
+            usage();
+            return false;
+        }
+
+        for (; !iter.done(); iter.next())
+        {
+            const char *arg = iter.query();
+            if (*arg != '-')
+            {
+                if (optSource.isEmpty())
+                    optSource.set(arg);
+                else if (optService.isEmpty())
+                    optService.set(arg);
+                else if (optMethod.isEmpty())
+                    optMethod.set(arg);
+                else if (diffTemplatePath.isEmpty())
+                    diffTemplatePath.set(arg);
+                else
+                {
+                    fprintf(stderr, "\nunrecognized argument: %s\n", arg);
+                    usage();
+                    return false;
+                }
+            }
+            else
+            {
+                if (iter.matchOption(optXsltPath, ESDLOPT_XSLT_PATH))
+                    continue;
+                if (EsdlConvertCmd::parseCommandLineOption(iter))
+                    continue;
+                if (EsdlConvertCmd::matchCommandLineOption(iter, true)!=EsdlCmdOptionMatch)
+                    return false;
+            }
+        }
+
+        return true;
+    }
+
+    virtual bool finalizeOptions(IProperties *globals)
+    {
+        if (optSource.isEmpty())
+        {
+            usage();
+            throw( MakeStringException(0, "\nError: Path to ESDL source file required\n"));
+        }
+
+        if( optService.isEmpty() )
+        {
+            usage();
+            throw( MakeStringException(0, "\nAn ESDL service name must be specified\n") );
+        }
+
+        if( optMethod.isEmpty() )
+        {
+            usage();
+            throw( MakeStringException(0, "\nAn ESDL method name must be specified\n") );
+        }
+
+        if (!diffTemplatePath.length())
+        {
+            usage();
+            throw( MakeStringException(0, "\nA differencing template name must be provided\n") );
+        }
+
+        if (optXsltPath.isEmpty())
+        {
+            StringBuffer tmp;
+            optXsltPath.set(getEsdlCmdComponentFilesPath(tmp));
+        }
+
+        diffTemplateContent.loadFile(diffTemplatePath);
+        cmdHelper.verbose = optVerbose;
+
+        return true;
+    }
+
+    class XpathTrack
```

Comment:
I may move this class somewhere even more common later... want to make it a bit more general though, after I've used it a bit more.